---
title: "QCAT_Meta"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{QCAT_Meta}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction
Microbiomics are one of the promising genomic research areas because microbes play
significant roles in human health. Types and abundances of microbes are also helpful indicators of
detecting human diseases and treatment effects. There have already existed lots of methods to test the
association between the covariates and microbial communities, such as the Dirichlet-Multinomial (DM)
test and Quasi-Conditional Association tests. However,
the limitation possesses in the power of such tests because of the small sample size of microbiome
research; the high background variation of microbes among different samples and meta-analysis is
beneficial to solve these problems.

In this R package, *MicroMeta*, I will implement both fixed effect
meta-analysis methods, for example, the FE-SKAT test , and random effect meta-analysis
methods such as Het-SKAT and RE-SKAT test based on the summary statistics generated
by the QCAT method. Considering that excessive zero often be observed with microbiome data, I
also employ generalized estimating equation (GEE) method to deal with the zero part of two part 
model which is based on the assumption that the data have a probability mass at zero and a response 
of positive values.

If the taxonomic hierarchy is provided, the tests will be performed on the
taxonomic tree to localize the covariate-associated lineages. If not, the tests will be performed using all
the taxon counts.

## Functionality

`MicroMeta` include two core function: `QCAT_Meta` and `QCAT_GEE_Meta`, both of them provide various arguments for users to customize the fit: we introduce some commonly used arguments here. (For more information, type `?MicroMeta`.)

`QCAT_Meta` 

* `OTU` is a list of matrices which contains OTU counts. For each matrices, 
its row corresponds to a sample and each column corresponds to an OTU or a taxa. Column name is mandatory and better the same across all the matrices. 

* `X` is a list of matrices. Each matrix contains covariates for the QCAT_Meta test with each column pertains to one variable (pertains to the covariate of interest or the confounders). The number of elements of of X and OTU must be the same. The column number of each matrix in this list must be the same.

* `X.index` is a vector indicate the columns in X for the covariate(s) of interest.

* `Tax` a matrix define the taxonomy ranks with each row corresponds to an OTU or a taxa and each column corresponds to a rank (start from the higher taxonomic level). Row name is mandatory and should be consistent with the column name of the OTU table,  Column name should be formated as "Rank1", "Rank2 ()"... etc. If provided, tests will be performed for lineages based on the taxonomic rank. The output contains P-values for all lineages; a list of significant lineages controlling the false discovery rate (based on resampling p-value if resampling test was performed); If not provided, one test will be performed with all the OTUs and one p-value will be output. The 
lowest rank in the Tax matrix must cover all the taxas appear in the OTU table list. 

* `Method` Meta-analysis method to be used in this test. Including fixed effect methods such as FE-MV test and FE-VC test and random effect methods like Het-SKAT and RE-SKAT test.

*`min.depth` is a number, if this parameter is set, test will only be perfomed on samples with depths greater or equal to this value.

*`n.perm` is a number, perform asymptotic test is n.perm is null, otherwise perform resampling tests using the specified number of resamplings. The random effect methods are mandatory to perform resampling tests. 

`fdr.alpha`  false discovery rate for multiple tests on the lineages.

`QCAT_GEE_Meta`

* `OTU` same as before. 

* `X` a list of matrices contains covariates for the zero-part test with each column pertains to one variable (pertains to the covariate of interest or the confounders). The number of elements of Z and OTU must be the same. The column number of each matrix in this list must be the same.

* `X.index` isa vector indicate the columns in X for the covariate(s) of interest.

* `Tax` same as before. 

* `Method` same as before.

* `min.depth` same as before.

* `n.perm` same as before. 

* `fdr.alpha`  same as before.

## Quick Start

#### If you have not installed ggtree, please run this chuck of code 
```{r, eval = F}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggtree")
```

#### Library required packages
```{r setup, message=FALSE, warning=FALSE}
library(MicroMeta)
library(dplyr)
library(data.tree)
library(ape)
library(ggtree)
library(ggplot2)
library(stringr)
library(HMP)
library(MASS)
library(geepack)
library(CompQuadForm)
```

## Load required data
```{r}
data(meta)
data(count.genus)
data(tax)
```

#### Data Processing
```{r}
## Split the OTU count table into five studies according to their data sources 
sample.id <- rownames(count.genus)
sample.id <- str_extract(sample.id,'^[A-Z]+') # extract their data source
sample.category <- unique(sample.id) # categories of data sources

##  split the samples according to their data source
otu.tab.lst <- split.data.frame(count.genus, sample.id)

rename.mat <- function(x,col,names){
  a <- matrix(x,ncol=col,byrow = F)
  colnames(a) = names
  return(a)
}

## For each matrix in the list, its columns names must be the scientific name of the genus (genus level)
otu.tab.lst <- lapply(otu.tab.lst, rename.mat, col = ncol(count.genus), names=colnames(count.genus))



## Find the column indexes for each otu.table which has more than 90% zero observations
.Remove.ind <- function(otu.tab.lst, ratio){
  n = length(otu.tab.lst)
  rev.ind.lst = NULL
  for(i in 1:n)
  {
    Y <- otu.tab.lst[[i]]
    Y0 = Y
    Y0[Y>0] = 0
    Y0[Y==0] = 1
    rev.ind.lst[[i]] <-which(colSums(Y0)/nrow(Y0)>ratio)
  }
  return (rev.ind.lst)
}

## Denote the genus that has more than 90% zero observations
remove.index.lst  <- .Remove.ind(otu.tab.lst, 0.9)

## Remove the indexes for each OTU table in the list
.Remove.col  <- function(otu.tab.lst,rev.ind.lst){
  if(length(rev.ind.lst) !=  length(otu.tab.lst))
  {
    stop("the number of elements of otu table list and remove index list should be the same")
  }
  else
  {
    n = length(otu.tab.lst)
    for(i in 1:n){
      rev.ind = rev.ind.lst[[i]]
      if(length(rev.ind) == 0){
        next
      }
      otu.tab.lst[[i]] = otu.tab.lst[[i]][ ,-rev.ind, drop = FALSE]
    }
  }
  return(otu.tab.lst)
}

otu.tab.lst  <- .Remove.col(otu.tab.lst,remove.index.lst)

## Rename the columns for Tax 
## Kingdom -> Rank1, Phylum -> Rank5, Class -> Rank4, ..., Genus -> Rank1
Tax <- tax[2:7] # Extract the Tax information needed
Tax <- Tax[!duplicated(Tax$genus),]
colnames(Tax) <- paste("Rank",c(1:ncol(Tax)),sep="")

## The row names for the Tax are mandatory and should cover all the taxa include in the origin OTU table (count.gene)
rownames(Tax) <- Tax$Rank6

## Construct the design matrix (Case group:1; Control Group: 0). Can add more covariates of interest like age, gender, etc. 
X <- rep(1,nrow(count.genus))
X[meta$Group == 'CTR'] <- 0
X <- split(X,sample.id)
X.index = 1
```

#### Tranform the Tax data into tree format
```{r}
Tax.df <- as.data.frame(Tax) #Tax is the taxonomy table
Tax.df$pathString <- paste(Tax.df$Rank1, Tax.df$Rank2, Tax.df$Rank3, Tax.df$Rank4, Tax.df$Rank5, Tax.df$Rank6, sep = "|") # taxonomic rank
Tax.tree <- as.Node(Tax.df, pathDelimiter = "|")
tree.node.name <- Tax.tree$Get("name", traversal = "level")
# tree.node.pval <- rep(NA, length(tree.node.name))
# names(tree.node.pval) <- tree.node.name
# tree.node.pval[colnames(one_part_VC$lineage.pval)] <- unname(one_part_VC$lineage.pval)
# tree.node.pval[!names(tree.node.pval)%in%colnames(one_part_VC$lineage.pval)] <- 1
# Tax.tree$Set(rawp = tree.node.pval, traversal = "level")
tree.vis <- as.phylo(Tax.tree)
```

## QCAT Model 

#### Fixed-effect Meta-analysis Method (Asymptotitc)
```{r}
## Asymptotic p-value
Method = "FE-VC"
one_part_VC <- QCAT_Meta(otu.tab.lst, X , 1, Tax = Tax, Method = Method, n.perm = NULL)
VC_sig_lineage <- one_part_VC$sig.lineage
VC_sig_lineage <- str_replace_all(VC_sig_lineage," ", "_")
#VC_sig_lineage <- VC_sig_lineage[VC_sig_lineage != "2_Bacteria"]
tip_num = length(tree.vis$tip.label)

# the index of the significant lineage
VC_sig_lineage <- unlist(lapply(1:length(VC_sig_lineage),function(X){tip_num+which(tree.vis$node.label == VC_sig_lineage[X])}))
print(one_part_VC$sig.lineage) 
```

#### Visualization

```{r fig.width= 4, fig.height= 4, fig.align='center', dpi=450}
p <- ggtree(tree.vis, layout='circular') +
  geom_hilight(node=137, fill="yellow", alpha= 0.6) +
  geom_hilight(node=237, fill="steelblue", alpha= 0.2) +
  geom_cladelab(node=137, label="Lactobacillales", align=TRUE,  angle=-55,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 1, fontsize=2) +
  geom_cladelab(node=237, label="Euryarchaeota", align=TRUE,  angle=-65,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = -0.5 , fontsize=4) +
  geom_point2(aes(subset=(node %in% VC_sig_lineage)),shape=21,  size=3, fill='red')

p + ggtitle("Significant lineage for FE-VC test (Asympotitc p-value)")
```

#### Random-effect Meta-analysis Method (Permutation)
```{r}
## Asymptotic p-value
Method = "Het-SKAT"
one_part_Het <- QCAT_Meta(otu.tab.lst, X , 1, Tax = Tax, Method = Method, n.perm = 450)
Het_sig_lineage <- one_part_Het$sig.lineage
Het_sig_lineage <- str_replace_all(Het_sig_lineage," ", "_")
#Het_sig_lineage <- Het_sig_lineage[Het_sig_lineage != "2_Bacteria"]

# the index of the significant lineage
Het_sig_lineage <- unlist(lapply(1:length(Het_sig_lineage),function(X){tip_num+which(tree.vis$node.label == Het_sig_lineage[X])}))
print(one_part_Het$sig.lineage) 
```

#### Visualization

```{r fig.width= 4, fig.height= 4, fig.align='center', dpi=450}
p <- ggtree(tree.vis, layout='circular') +
  geom_hilight(node=174, fill="darkgreen", alpha= 0.6) +
  geom_hilight(node=154, fill="darkgreen", alpha= 0.6) +
  geom_cladelab(node=174, label="Peptoniphilaceae", align=TRUE,  angle=85,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = -0.5, fontsize=2) +
  geom_cladelab(node=154, label="Peptostreptococcaceae", align=TRUE,  angle=15,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 1, fontsize=2) +
  geom_point2(aes(subset=(node %in% Het_sig_lineage)),shape=21,  size=3, fill='red')

p + ggtitle("Significant lineage for Het-SKAT test (Permutation p-value)")
```
## QCAT_GEE Model 

## Rarefy the data

```{r}
## Rarefy
Rarefy <- function (otu.tab, depth = min(rowSums(otu.tab)),seed = 1145){
  # Rarefaction function: downsample to equal depth
  #	
  # Args:
  #		otu.tab: OTU count table, row - n sample, column - q OTU
  #		depth: required sequencing depth 
  #
  # Returns:
  # 	otu.tab.rff: Rarefied OTU table
  #		discard: labels of discarded samples
  
  otu.tab <- as.matrix(otu.tab)
  ind <- (rowSums(otu.tab) < depth)
  sam.discard <- rownames(otu.tab)[ind]
  otu.tab <- otu.tab[!ind, ]
  
  rarefy <- function(x, depth){
    set.seed(seed)
    y <- sample(rep(1:length(x), x), depth)
    y.tab <- table(y)
    z <- numeric(length(x))
    z[as.numeric(names(y.tab))] <- y.tab
    z
  }
  otu.tab.rff <- t(apply(otu.tab, 1, rarefy, depth))
  rownames(otu.tab.rff) <- rownames(otu.tab)
  colnames(otu.tab.rff) <- colnames(otu.tab)
  return(list(otu.tab.rff=otu.tab.rff, discard=sam.discard))
}

otu.tab.lst <- lapply(otu.tab.lst,function(X){Rarefy(X)$otu.tab.rff})


```

#### Random-effect Meta-analysis Method (Asymptotitc)

```{r}
## Permutation p-value
zero_part_RE <- QCAT_GEE_Meta(otu.tab.lst, X , 1, Tax = Tax, Method = "RE-SKAT", n.perm = 300)
RE_sig_lineage <- zero_part_RE$sig.lineage
RE_sig_lineage <- str_replace_all(RE_sig_lineage," ", "_")
#RE_sig_lineage <- RE_sig_lineage[RE_sig_lineage != "2_Bacteria"]
tip_num = length(tree.vis$tip.label)

# the index of the significant lineage
RE_sig_lineage <- unlist(lapply(1:length(RE_sig_lineage),function(X){tip_num+which(tree.vis$node.label == RE_sig_lineage[X])}))
print(zero_part_RE$sig.lineage) 
```

```{r fig.width= 4, fig.height= 4, fig.align='center', dpi=450}

## Visualize  some of the significant lineage                                        
p <- ggtree(tree.vis, layout='circular') +
  geom_hilight(node=158, fill="steelblue", alpha= 0.6) +
  geom_hilight(node=174, fill="steelblue", alpha= 0.6) +
  geom_hilight(node=216, fill="yellow", alpha= 0.6) +
  geom_hilight(node=154, fill="yellow", alpha= 0.6) +
  geom_hilight(node=214, fill="darkgreen", alpha= 0.6) +
  geom_hilight(node=190, fill="darkgreen", alpha= 0.6) +
  geom_hilight(node=178, fill="steelblue", alpha= 0.2, type="rect") +
  geom_hilight(node=237, fill="steelblue", alpha= 0.2) +
  geom_cladelab(node=158, label="Clostridiaceae", align=TRUE,  angle=-55,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 0.75 , fontsize=2)+
  geom_cladelab(node=174, label="Peptoniphilaceae", align=TRUE,  angle=0,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 0.75 , fontsize=2)+
  geom_cladelab(node=216, label="Porphyromonadaceae", align=TRUE,  angle=-67,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = -0.5 , fontsize=2) +
  geom_cladelab(node=154, label="Peptostreptococcaceae", align=TRUE,  angle= -75,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 1, fontsize=2) +
  geom_cladelab(node=214, label="Prevotellaceae", align=TRUE,  angle=-77,
                offset = 9.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 0, fontsize=2) +
  geom_cladelab(node=190, label="Sutterellaceae", align=TRUE,  angle=45,
                offset = 9.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = 0, fontsize=2) +
  geom_cladelab(node=178, label="Proteobacteria", align=TRUE, angle = 60,
                offset = 15.0, textcolor='black', barcolor='black', hjust= 0.5, vjust= -0.5, fontsize=6) +
  geom_cladelab(node=237, label="Euryarchaeota", align=TRUE,  angle=-65,
                offset = 2.0, textcolor='black', barcolor='black', hjust= 0.5, vjust = -0.5 , fontsize=4) +
  geom_point2(aes(subset=(node %in% c(174,216,214,154,158,190,178,237,134))),shape=21,  size=3, fill='red')

p + ggtitle("Significant lineage for zero part RE-SKAt test (Permutation p-value)")
```
